src/DomainGuest.sol

```diff
@@ -232,43 +235,49 @@ abstract contract DomainGuest {
         emit Release(burned);
     }
 
-    /// @notice Push surplus (or deficit) to the host dss
+    /// @notice Push surplus to the host dss
     /// @dev Should be run by keeper on a regular schedule
-    function _push() internal returns (uint256 _rid, int256 _surplus) {
+    function _surplus() internal returns (uint256 _rid, uint256 wad) {
         require(live == 1, "DomainGuest/not-live");
 
         _rid = rid++;
 
         uint256 _dai = vat.dai(address(this));
         uint256 _sin = vat.sin(address(this));
-        if (_dai >= _sin + dust) {
-            // We have a surplus
-            if (_sin > 0) vat.heal(_sin);
+        require(_dai > _sin, "DomainGuest/non-surplus");
+        unchecked { wad = (_dai - _sin) / RAY; } // Round against this contract for surplus
+        require(wad >= dust, "DomainGuest/dust");
 
-            uint256 wad = (_dai - _sin) / RAY;    // Round against this contract for surplus
+        // Burn the DAI and unload on the other side
+        vat.swell(address(this), -_int256(wad * RAY));
 
-            // Burn the DAI and unload on the other side
-            vat.swell(address(this), -_int256(wad * RAY));
-            _surplus = _int256(wad);
+        emit Surplus(wad);
+    }
 
-            emit Push(int256(wad));
-        } else if (_sin >= _dai + dust) {
-            // We have a deficit
-            if (_dai > 0) vat.heal(_dai);
+    /// @notice Push deficit to the host dss
+    /// @dev Should be run by keeper on a regular schedule
+    function _deficit() internal returns (uint256 _rid, uint256 wad) {
+        require(live == 1, "DomainGuest/not-live");
 
-            int256 deficit = -_int256(_divup(_sin - _dai, RAY));    // Round up to overcharge for deficit
-            _surplus = deficit;
+        _rid = rid++;
 
-            emit Push(deficit);
-        } else {
-            revert("DomainGuest/dust");
-        }
+        uint256 _dai  = vat.dai(address(this));
+        uint256 _dsin = dsin;
+        uint256 _sin  = vat.sin(address(this)) - _dsin * RAY;
```

Comment:
This can revert with a non-descript arithmetic error if the sin has decreased below `dsin` which can happen for any number of reasons (revenue comes in, permissionless repay, etc). I believe this is what @olivdb was referring to above. I would add an explicit error with a better error message.

Reply:
Done.